### DISCUSSION TITLE
Different dynamic icons for a single application.

### DISCUSSION THREAD

**Discussion by adam-golab** [vercel/next.js#52921] [App Router] [answered]
Hi everyone,

I'm working on a site builder project and encountered odd issues when tried to set the correct icon in the head section. Prior to the next 13 version everything worked just fine as I could use `getServerSideProps` function get the correct URL of icon and render it inside of the `<Head>` element. I used the server side rendering approach. Because it can be controlled in the admin panel I can't just use static files for setting the correct icon. 

Now, I wanted to use the next 13 App Router and Server Components to achieve the same results but from what I see in the documentation, adding the icon has been automatized and there is no way to control that in the same way as previously. 

What I'm trying right now is the usage of the `icon.tsx` file in the root directory. So, I created a following file:
```ts
import { headers } from 'next/headers';

export const runtime = 'nodejs';

export const size = {
  width: 32,
  height: 32,
};
export const contentType = 'image/jpeg';

const Icon = async (): Promise<ArrayBuffer> => {
  const host = headers().get('host') || DEFAULT_HOSTNAME;

  const configuration = await fetch(`${API_URL}/configuration`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      'remote-host': host,
    },
  }).then((res) => res.json());

  const image = await fetch(`${API_URL}/icons/${configuration.faviconFilename}`).then((res) => res.arrayBuffer());

  return image;
};

export default Icon;
```

The `link` element is added correctly to the `head` section but I'm getting `500 Internal Server error` on the requests for the icon. 

What am I doing wrong? Maybe there is another better way to handle adding the icon to the app?

**Comment by icyJoseph** [comment, marked as answer]
I am not sure why it is not working like that, but... if you do:

```ts
const Icon = async () => {
  const host = headers().get('host') || DEFAULT_HOSTNAME;

  const configuration = await fetch(`${API_URL}/configuration`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      'remote-host': host,
    },
  }).then((res) => res.json());

  return fetch(`${API_URL}/icons/${configuration.faviconFilename}`)
};

```

What do you get?

**Reply by adam-golab** [reply]
@icyJoseph this seems to work. Thanks! I don't know why I haven't tested that.

So, to summarize it, it looks like the `Blob`, `ArrayBuffer` or `ReadableStream` is not supported there. It needs to be just `Response`

**Reply by icyJoseph** [reply]
Though the documentation says that those are supported 👀
